From 6aba834ae4d3c1f48727cd075317916c14681598 Mon Sep 17 00:00:00 2001
From: Dan Liang <dan.liang@intel.com>
Date: Tue, 30 Apr 2013 14:26:16 +0800
Subject: libstagefright: adjust audio seek for mkv container

BZ: 57484

The original seeking method can only deal with the case
that the SimpleBlocks of both audio and video are packed
in the head of a cluster. The testing clip is quite
special, as the key frame for video is located at the end
of a cluster, at 10000ms, while the first SimpleBlock of
audio in this cluster starts at 5013ms. So video is stopped
untill audio catches up.

After modification, audio will advance to the nearest point
as video has seeked to.

Category: aosp improvement
Domain: Video.Media-container
Origin: internal
Upstream-Candidate: yes

Change-Id: I48f6bc2a9e78256a08f023052997aeb19ab79226
Orig-Change-Id: Ie9c62df10f2f64004ffde9e36bf8a67608246fff
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 .../libstagefright/matroska/MatroskaExtractor.cpp  |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)
 mode change 100644 => 100755 media/libstagefright/matroska/MatroskaExtractor.cpp

diff --git a/media/libstagefright/matroska/MatroskaExtractor.cpp b/media/libstagefright/matroska/MatroskaExtractor.cpp
old mode 100644
new mode 100755
index 18be213..c450db3
--- a/media/libstagefright/matroska/MatroskaExtractor.cpp
+++ b/media/libstagefright/matroska/MatroskaExtractor.cpp
@@ -410,8 +410,9 @@ void BlockIterator::seek(
 
         if (eos()) break;
 
-        if (isAudio || block()->IsKey()) {
-            // Accept the first key frame
+        if((!isAudio && block()->IsKey()) ||
+            // audio track is allowed to be 10ms earlier than video track
+            (isAudio && blockTimeUs() + 10000LL >= seekTimeUs)) {
             *actualFrameTimeUs = (block()->GetTime(mCluster) + 500LL) / 1000LL;
             ALOGV("Requested seek point: %lld actual: %lld",
                   seekTimeUs, actualFrameTimeUs);
-- 
1.7.4.1

